FRAuth/FRAuth/DeviceBinding/ApplicationPinDeviceAuthenticator.swift

```diff
@@ -60,4 +70,62 @@ open class ApplicationPinDeviceAuthenticator: DeviceAuthenticator, CryptoAware {
     open func setPrompt(_ prompt: Prompt) {
         self.prompt = prompt
     }
+    
+    
+    // Default implemention
+    /// Sign the challenge sent from the server and generate signed JWT
+    /// - Parameter userKey: user Information
+    /// - Parameter challenge: challenge received from server
+    /// - Parameter expiration: experation Date of jws
+    /// - Returns: compact serialized jws
+    public func sign(userKey: UserKey, challenge: String, expiration: Date) throws -> String {
+        guard let prompt = prompt else {
+            throw DeviceBindingStatus.unsupported(errorMessage: "Cannot retrive keys, missing prompt")
+        }
+        
+        guard let pin = collectPin(prompt: prompt) else {
+            throw DeviceBindingStatus.abort
+        }
+        
+        guard let keyStoreKey = CryptoKey.getSecureKey(keyAlias: userKey.keyAlias, pin: pin) else {
+            throw DeviceBindingStatus.unsupported(errorMessage: "Cannot read the private key")
+        }
+        let algorithm = SignatureAlgorithm.ES256
+        
+        //create header
+        var header = JWSHeader(algorithm: algorithm)
+        header.kid = userKey.kid
+        header.typ = "JWS"
+        
+        //create payload
+        let params: [String: Any] = ["sub": userKey.userId, "challenge": challenge, "exp": (Int(expiration.timeIntervalSince1970))]
+        let message = try JSONSerialization.data(withJSONObject: params, options: [])
+        let payload = Payload(message)
+        
+        //create signer
+        guard let signer = Signer(signingAlgorithm: algorithm, key: keyStoreKey) else {
+            throw DeviceBindingStatus.unsupported(errorMessage: "Cannot create a signer for jws")
+        }
+        
+        //create jws
+        let jws = try JWS(header: header, payload: payload, signer: signer)
+        
+        return jws.compactSerializedString
+    }
+    
+    
+    func collectPin(prompt: Prompt) -> String? {
+        // Use DispatchGroup to wait for pinCollector to finish collecting the pin before continuing
+        let group = DispatchGroup()
+        var pin: String?
+        
+        group.enter()
+        pinCollector.collectPin(prompt: prompt, completion: { collectedPin in
+            pin = collectedPin
```

Comment:
So when it returns nil, it considers canceling?

Reply:
Yes Andy, as the Ok button is disabled when the pin is empty.